### ISSUE TITLE
Proposal for 3.x: per-partition control flow

### ISSUE THREAD

**Issue by thomaslee** [SOHU-Co/kafka-node#994]
I'd love to be able to tell kafka-node to give me _at most_ one message at a time per partition so I could do something like the following:

```
const consumer = new ConsumerGroup(...)
consumer.on('message', msg => {
  consumer.pause(msg.partition)
  someSlowOperation(msg).then(() => consumer.resume(msg.partition))
})
```

A poor-man's implementation of this using 2.6.x might simply call pause() for each message, but
a) that wouldn't currently prevent the rest of the batch being processed, so even though you call pause() you still see every single message in the batch until you hit a `'done'` indicating the end of a batch; and b) even if it worked that way, it constrains parallelism to an unsatisfactory degree. This leads to us needing to reach for relatively complicated solutions like `async.queue`, which feels a bit unnecessary.

The 2.x model also feels like a bit of a leaky abstraction: folks need to know about `'message'` and `'done'` and all the batching weirdness we need to keep performance snappy -- it seems like all they should need to think about is `'message'` & then pull levers on the consumer to influence the flow of message processing.

(Obviously as implemented today the above would have performance implications because resume() would immediately kick off a new fetch(). Just trying to convey the general idea.)

**Comment by aikar** [comment]
I've been discussing with hyperlink today on this. My solution is not really like you said, but as I mentioned here: https://github.com/SOHU-Co/kafka-node/issues/975

We need to support promise based message handlers.

The easy way is to use async-events to have an Async Event Emitter.
Then, the ConsumerGroup should queue up messages into a Map<string, Map<number, Array<Message>>> queue.

Then we only pop the next message off of the queue once the promise resolves for that topic/partition, and only commit it at that point.

However this API must also support manual commit, so something like:

```javascript
on('message', async (msg, transaction) => {
    const preparedJob = await PreparedJob.create(msg);
    // optionally: transaction.start();
    try {
        await transaction.commit(false); // force flag. when false, commits will batch upon commit interval as they do in ConsumerGroupStream, allowing multiple topic/partition commits to batched together.
        job.execute();
    } catch (e) {
        job.cancel();
        throw e;
    }
});
```

as we also need to support Kafka Transactions too.

I'm in favor of dropping ConsumerGroupStream completely, and then offer a "Message Handler" that can pipe into a readable, and then Stream users would use ConsumerGroup, and do 
```javascript
consumer.on('message', new MessageStreamProcessor(consumer).pipe(new MyProcessor()));
```

**Comment by thomaslee** [comment]
Right, my proposal was ugly -- was just trying to convey something that is currently quite hard to achieve with kafka-node assuming no grand redesign of the API was planned for 3.x. Maybe that's a bad assumption.

I think we actually have similar ideas of what this looks like in an ideal world. The API that we've been using in our Node 8.x wrapper around kafka-node looks something like this:

```javascript
const consumer = new Consumer(options)

consumer.on('error', err => {
  //
  // Log anything that bubbles up.
  //
})

//
// The callback passed to consume() is called per message.
//
consumer.consume(async msg => {
  await processAsync(msg)

  //
  // Returns a promise that resolves when the message is confirmed committed.
  //
  msg.commit()
})

process.on('SIGTERM', () => consumer.close().then(() => process.exit(0))
```

The callback passed to `consume` can be called multiple times in parallel: as many messages can be in-flight in parallel as there are topic partitions assigned to the consumer.

But I'm not too fussed about the finer details of the API itself -- so long as it's _possible_ to use Kafka the way it was intended (consume -> process -> commit -> consume ...). The existing APIs come close, but there are obviously gaps. It's difficult to get the correct semantics without subclassing kafka-node classes. If the async emitter stuff is an easier middle ground that we can shoot for in 3.x, great.
